### ISSUE TITLE
Support SearchView

### ISSUE THREAD

**Issue by eburnette** [JakeWharton/ActionBarSherlock#70]
I'm trying to port a program that uses this menu item:

```
<item
    android:id="@+id/menu_search"
    android:title="@string/search_action"
    android:icon="@android:drawable/ic_menu_search"
    android:actionViewClass="android.widget.SearchView"
    android:showAsAction="always" />
```

but when I run it I get:

09-30 17:13:26.974: ERROR/AndroidRuntime(4085): Caused by: android.view.InflateException: java.lang.ClassNotFoundException: android.widget.SearchView
09-30 17:13:26.974: ERROR/AndroidRuntime(4085):     at android.support.v4.view.MenuInflater$ActionBarMenuState.setItem(MenuInflater.java:410)
09-30 17:13:26.974: ERROR/AndroidRuntime(4085):     at android.support.v4.view.MenuInflater$ActionBarMenuState.addItem(MenuInflater.java:419)
09-30 17:13:26.974: ERROR/AndroidRuntime(4085):     at android.support.v4.view.MenuInflater.parseMenu(MenuInflater.java:177)
09-30 17:13:26.974: ERROR/AndroidRuntime(4085):     at android.support.v4.view.MenuInflater.inflate(MenuInflater.java:103)

I couldn't find a compatible version of SearchView.

**Comment by JakeWharton** [comment]
I debated about whether or not this was "in scope" of this library a while back and ultimately ended up deciding that it was not.

For now, I would recommend supplying two menu XMLs in `menu` and `menu-v11`. The former would only specify the icon and the latter would include the search view. If memory serves the search view will not dispatch an event to `onOptionsItemSelected` so if that even is fired you can react accordingly (likely dispatching [a search dialog](http://developer.android.com/guide/topics/search/search-dialog.html)).

Once the Ice Cream Sandwich sources are released I will either include it or release it as a standalone library.

If you're feeling ambitious, I can provide the decompiled smali and associated resources from Android 3.2 and you can take a hack at reconstructed the Java...

**Comment by doc-rj** [comment]
Hi Jake, I'm also interested in using a SearchView widget with ActionBarSherlock.  The Ice Cream Sandwich source was released a few days ago.  Are you still planning to add support for SearchView?  Thanks!

**Comment by JakeWharton** [comment]
Yeah, but it's a bit down on "the list", however. ActionBarSherlock 4.0, scrollable tab indicators for ViewPagerIndicator, and a secret experimental project are above it. If I get bored I'll probably give it a go though.

**Comment by JakeWharton** [comment]
Started in the `search-view` branch. Lots of internal code related to querying the [`SearchManager`](http://developer.android.com/reference/android/app/SearchManager.html) which may need to be stripped out.

**Comment by bostone** [comment]
I tried to get search-view branch and got tons of compilation errors related to the resources I think

**Comment by klnusbaum** [comment]
Just thought I'd throw my two cents in and say this is something I'd really like to have available. Is there anyway I can help get this feature implemented?

**Comment by DylanSale** [comment]
This would be very handy for me also. Let me know if I can help.

**Comment by bostone** [comment]
Just checking if there's any progress on this?

**Comment by yaroslav-v** [comment]
+1 for this feature. This'll be very helpful.

**Comment by nschwermann** [comment]
very annoying and my guess is they are just being lazy in using all these internal widgets and resource files. Google confuses me with the support library sometimes.

**Comment by imort** [comment]
Can this be solved with SeachViewCompat from support package v4?

**Comment by navarr** [comment]
@imort: It looks like it just presents a constructor for creating a SearchView on platforms it exists on, and returning null otherwise - instead of trying to figure out if it exists or not.

**Comment by JakeWharton** [comment]
A preliminary, `SearchableInfo`-free version of this is incubating on the `dev` branch. Please try it out.

**Comment by inazaruk** [comment]
I've looked at this issue 2 days ago and decided to try and solve it. Didn't know you were working on it too.

Finished yesterday late. My port takes a slightly different approach than yours (though I still tried to stick with ABS way of porting it). You can take a look at it at: https://github.com/inazaruk/abs-search-view.

From quick look through your version, I'd say that main difference is that my version falls back to native `android.widget.SearchView` on API levels 11 and higher. This is because I wasn't able to support all the key features related to `SearchableInfo`, but still wanted to enable those features on platforms that support it.

**Comment by b-ashtiani** [comment]
I'm new to Android so I appreciate it if Jake or anyone else could clarify what the lack of SearchableInfo entails? In other words which features of the vanilla searchView are unsupported? I did review the code but this is not immediately apparent from the code.

**Comment by navarr** [comment]
@b3hrooz I don't entirely understand it myself, but what it seems to entail is most if not all of the search functionality, including the ability to auto-complete etc. 

http://developer.android.com/guide/topics/search/search-dialog.html#SearchableConfiguration

**Comment by b-ashtiani** [comment]
@navarr I have been reading the dev guide and it seemed to me that the vanilla search widget and dialog are heavily dependent on searchableInfo (but I could be wrong). What's confusing is Jake's comment on the new feature reads "Add SearchView widget which supports a very large subset of the API." and looking at the code I even see SuggestionAdaptors and such... I am trying the code right now but someone with experience using the vanilla searchView widget can probably clarify this much sooner than I can.

Basically what I want to know is whether this widget has autocomplete, suggestions, and voice input to set it apart from a regular textView.

**Comment by JakeWharton** [comment]
Autocomplete and suggestions are the same thing and are provided by a `Cursor`. This is not currently in the example inside the 'demos' sample but would make a great addition. Voice search is not available (I forget why it wouldn't work).

**Comment by b-ashtiani** [comment]
Thanks Jake, that's great news. I'm guessing the lack of voice search is not the biggest problem since we can always just have it as an action provider and feed the results back into the search widget.

**Comment by inazaruk** [comment]
Voice search is configured through SearchableInfo, and because of lack of support of SearchableInfo there is probably no support for voice too. At least this was my understanding when I looked through code.

**Comment by gabrielittner** [comment]
Do you plan to add SearchableInfo? If not could you at least add SuggestionsAdapter and com.android.internal.R.layout.search_dropdown_item_icons_2line so we can add suggestions that are closer to the platform?
If there is anything I can do to help you just let me know.

**Comment by Shusshu** [comment]
+1 replying to get the notifications from this topic as it will be interesting for me in the near future

**Comment by lujop** [comment]
I was previously using setSearchableInfo and it's not clear to me how to workarround that.
My first big doubt is why SuggestionsAdapter is package protected? It seems that it's usable and makes the job.  I'm ignoring something?

What's the best and recommended way to migrate from searchableinfo to be able to use the SearchView in ActionBarSherlock?

**Comment by JakeWharton** [comment]
It's package scoped and annotated with `@hide` in Android as well.

**Comment by lujop** [comment]
I see Jake. But I suppose it's intended to be use by the searchableinfo, isn't it? With actual ABS code it seems that is never used.

Also I've been playing a little with it and I think that there are some attributes not set in themes:
?attr/searchResultListItemHeight, ?attr/searchViewEditQuery ,searchViewEditQueryBackground

**Comment by dylanreid** [comment]
Really looking forward to this

**Comment by bostone** [comment]
Actually - it maybe already there. Not sure what is the state but I'm using
zip form the dev branch and SearchView is already there

**Comment by JakeWharton** [comment]
It's in the `dev` branch. It's not done. I have to re-add `SearchableInfo` support.

**Comment by bostone** [comment]
Good enough to start coding using this as "things to come". Thank you Jake - your stuff rocks!

**Comment by lujop** [comment]
That's nice Jake. Then will we be able to use the SearchableInfo as in oficial SearchView?

**Comment by yincrash** [comment]
I'm on a22e243de534e235663bd1081e5beaeec7bb0fe5 of the dev branch (the latest commit at time of posting), and i tried adding a recent suggestions provider, but it didn't work. The popup that shows suggestions never pops up. However, switching from the ABS SearchView to the android-16 SearchView immediately had working recent suggestions.

``` java
SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
((SearchView) menu.findItem(R.id.action_search).getActionView()).setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
```

is replaced with 

``` java
SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
android.widget.SearchView searchView = new android.widget.SearchView(getSupportActionBar().getThemedContext());
searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
menu.findItem(R.id.action_search).setActionView(searchView);
```

**Comment by JakeWharton** [comment]
Thanks, @yincrash. I will investigate.

**Comment by JakeWharton** [comment]
@yincrash I never got a chance to test this fully and I'm not going to block the 4.2 release for this. Can you do two things:
1. Create a new issue for this.
2. Create a pull request against the `dev` branch which adds an example to the 'known-bugs` sample which shows the issue.
